torch/_higher_order_ops/invoke_subgraph.py

```diff
@@ -253,54 +258,140 @@ def create_fw_bw_graph(subgraph, operands, grad_outputs=None):
+
+        # return signature is deliberately kept (*grads, *fw_outs). This
+        # simplifies partitioning work later on.
+        return pytree.tree_map(maybe_clone, tuple(grads + list(fw_outs)))
+
+    with suspend_functionalization(), disable_functional_mode():
+        with disable_proxy_modes_tracing():
+            joint_operands = [_from_fun(arg) for arg in joint_operands]
+            with contextlib.ExitStack() as stack:
+                stack.enter_context(
+                    torch._C._ForceDispatchKeyGuard(include_key_set, exclude_key_set),
+                )
+                with torch.enable_grad():
+                    return _maybe_reenter_make_fx(joint_fn)(*joint_operands)
 
 
 class InvokeSubgraphAutogradOp(torch.autograd.Function):
     """
-    This autograd function op is to stash the backward graph in the ctx while
-    running forward.
+    Saves the subgraph, i.e. original callable, in the forward method. And then
+    traces out a joint graph in the backward. This delaying of tracing in
+    backward, also called as lazy backward, ensures that the assumptions about
+    the grad_out strides and tensor-subclass-ness are already accounted for.
     """
 
     @staticmethod
     def forward(
         ctx,
-        fw_graph,
-        bw_graph,
+        subgraph,
         identifier,
-        num_fw_outs,
-        filter_tangent_info,
+        output_metadata,
         *operands,
     ):
-        ctx._fw_graph = fw_graph
-        ctx._bw_graph = bw_graph
+        # We want to delay the backward graph construction until the backward.
+        # So in forward, we just run the fw callable as is. And save all the
+        # information necessary to construct the backward graph in the ctx.
+        ctx._subgraph = subgraph
         ctx._identifier = identifier
-        ctx._num_fw_outs = num_fw_outs
-        ctx._filter_tangent_info = filter_tangent_info
+        ctx._output_metadata = output_metadata
+        # We snapshot the dispatch keys in forward for materializing the
+        # the bw_graph in backward.
+        ctx._fw_include_key_set = torch._C._dispatch_tls_local_include_set()
+        ctx._fw_exclude_key_set = torch._C._dispatch_tls_local_exclude_set()
+
+        save_tensors_and_symints_for_backward(ctx, operands)
 
         with torch._C._AutoDispatchBelowAutograd():
             out = invoke_subgraph(
-                fw_graph,
+                subgraph,
                 f"___forward_{identifier}",
                 operands,
             )
 
-        save_tensors_and_symints_for_backward(ctx, operands)
-
         # Check that None is at expected indexes.
         for idx, o in enumerate(out):
             if o is None:
-                assert idx in filter_tangent_info.indexes_with_none
+                assert idx in output_metadata.indexes_with_none
 
         return out
 
     @staticmethod
-    def backward(ctx, *grad_outs):
-        bw_graph = ctx._bw_graph
+    def backward(
```

Comment:
One compile time problem here is that earlier we were just doing tracing once and we were saving the fw and bw graph. But now, we are doing tracing in every `backward`. So, somehow we need to cache the bwd graph as well.

This might explain the increase of compile time from 56 second to 70 second.

Reply:
This is resolved. Added cache for the bwd graph.